predictive_test_selection/data_collector.py

```diff
@@ -0,0 +1,236 @@
+        self._compute_file_extensions()
+        self._count_modifications()
+
+    def _compute_changed_files(self, current_branch: str) -> None:
+        """
+        Finds the development branch that is the closest to this code change and sets the relevant attributes accordingly.
+        The distance metric used is the total number of files in the diff with the current branch
+        """
+        max_cardinality: Union[float, int] = float("inf")
+
+        for dev_branch in DEV_BRANCHES:
+            cmd = ["git", "diff", f"{dev_branch}...{current_branch}", "--name-only"]
+            changed_files = [line.strip() for line in subprocess.check_output(cmd).decode().split("\n") if line.strip()]
+
+            current_cardinality = len(changed_files)
+
+            if current_cardinality < max_cardinality:
+                max_cardinality = current_cardinality
+                self.dev_branch = dev_branch
+                self.changed_files = changed_files
+                self.file_cardinality = current_cardinality
+
+            if max_cardinality == 0:
+                raise AbortDataCollection(f"the code change is aligned with dev branch {dev_branch}.")
+
+    def _count_modifications(self) -> None:
+        """
+        Counts the number of modifications that have been made to the files involved in this change for each interval (in days)
+        specified in the past CHANGE_LOOK_BACK parameter.
+        """
+        modification_count: List[int] = []
+        for n_days_ago in CHANGE_LOOK_BACK:
+            acc: int = 0
+            for file in self.changed_files:
+                ps = subprocess.Popen(
+                    ["git", "log", f"--after='{n_days_ago} days ago'", "--format=oneline", f"{file}"],
+                    stdout=subprocess.PIPE,
+                )
+                try:
+                    acc += int(subprocess.check_output(["wc", "-l"], stdin=ps.stdout))
+                finally:
+                    ps.stdout.close()
+                    ps.wait()
+
+            modification_count.append(acc)
+        self.modification_count = modification_count
+
+    def _compute_file_extensions(self) -> None:
+        """
+        Computes the set of file extensions present in this code change
+        """
+        self.file_extensions = set([os.path.splitext(file)[1].replace(".", "") for file in self.changed_files])
+
+    def get_modifications_data_format(self) -> str:
+        """
+        Converts the modification_count (List[int]) to a condensed csv format.
+        e.g [1, 2, 3] becomes "1,2,3"
+        """
+        return str(self.modification_count).replace(" ", "")[1:-1]
+
+    def get_file_extensions_data_format(self) -> str:
+        """
+        Converts the file_extensions (Set[str]) to a condensed csv format.
+        e.g {'txt', 'md', 'py'} becomes 'txt,md,py'
+        """
+        return str(self.file_extensions).replace(" ", "").replace("'", "").replace('"', "")[1:-1]
```

Comment:
string manipulation to format structured data is something that scares me
```suggestion
                return ",".join(self.file_extensions)
```